plugins/ccgx/fu-ccgx-hpi-device.c

```diff
@@ -84,12 +612,151 @@ fu_ccgx_hpi_write_firmware (FuDevice *device,
 }
 
 static gboolean
-fu_ccgx_hpi_device_setup (FuDevice *device, GError **error)
+fu_ccgx_hpi_device_ensure_silicon_id (FuCcgxHpiDevice *self, GError **error)
 {
+	guint8 buf[2] = { 0x0 };
+	g_autofree gchar *instance_id = NULL;
+
+	if (!fu_ccgx_hpi_device_reg_read (self, CY_PD_SILICON_ID,
+					  buf, sizeof(buf), error)) {
+		g_prefix_error (error, "get silicon id error: ");
+		return FALSE;
+	}
+	if (!fu_common_read_uint16_safe (buf, sizeof(buf),
+					0x0, &self->silicon_id,
+					G_LITTLE_ENDIAN, error))
+		return FALSE;
+
+	/* add quirks */
+	instance_id = g_strdup_printf ("CCGX\\SID_%X", self->silicon_id);
+	fu_device_add_instance_id_full (FU_DEVICE (self),
+					instance_id,
+					FU_DEVICE_INSTANCE_FLAG_ONLY_QUIRKS);
+
+	/* sanity check */
+	if (self->flash_row_size == 0x0) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "Invalid row size for Instance ID: %s",
+			     instance_id);
+		return FALSE;
+	}
+
 	/* success */
 	return TRUE;
 }
 
+static gboolean
+fu_ccgx_hpi_device_setup (FuDevice *device, GError **error)
+{
+	FuCcgxHpiDevice *self = FU_CCGX_HPI_DEVICE (device);
+	CyI2CConfig i2c_config = { 0x0 };
+	guint32 hpi_event = 0;
+	guint8 mode = 0;
+	g_autofree gchar *instance_id = NULL;
+	g_autoptr(GError) error_local = NULL;
+
+	/* set the new config */
+	if (!fu_ccgx_hpi_device_get_i2c_config (self, &i2c_config, error)) {
+		g_prefix_error (error, "get config error: ");
+		return FALSE;
+	}
+	i2c_config.frequency = FU_CCGX_HPI_FREQ;
+	i2c_config.is_master = TRUE;
+	i2c_config.is_msb_first = TRUE;
+	if (!fu_ccgx_hpi_device_set_i2c_config (self, &i2c_config, error)) {
+		g_prefix_error (error, "set config error: ");
+		return FALSE;
+	}
+	if (!fu_ccgx_hpi_device_reg_read (self, CY_PD_REG_DEVICE_MODE_ADDR,
+					  &mode, 1, error)) {
+		g_prefix_error (error, "get device mode error: ");
+		return FALSE;
+	}
+	self->hpi_addrsz = mode & 0x80 ? 2 : 1;
+	self->num_ports = (mode >> 2) & 0x03 ? 2 : 1;
+	self->fw_mode = (FWMode) (mode & 0x03);
+
+	/* add extra instance ID */
+	instance_id = g_strdup_printf ("USB\\VID_%04X&PID_%04X&MODE_%s",
+				       fu_usb_device_get_vid (FU_USB_DEVICE (device)),
+				       fu_usb_device_get_pid (FU_USB_DEVICE (device)),
+				       fu_ccgx_fw_mode_to_string (self->fw_mode));
+	fu_device_add_instance_id (device, instance_id);
```

Comment:
Realistically do you see these 3 IDs as the ones being used to update FW and not the other ones?  I wonder if the "generic" IDs should actually be nuked in such case.  We have precedent in doing that in NVME with the Dell GUID's too.

Reply:
It's a good question. I'm not 100% sure how we're going to do the matching yet, I added this instance ID just for the custom UpdateMessage string. On this case, I think still having the other "generic" VID/PID ones are useful, as we can also do the CounterpartGuid stuff where we don't care which firmware is running.